metadata/metadata.go

```diff
@@ -44,9 +45,37 @@ type (
 	}
 )
 
+// metadataContext holds context for metadata serialization and deserialization.
+type metadataContext struct {
+	protocols map[multicodec.Code]func() Protocol
+}
+
+var Default metadataContext
+
+func init() {
+	Default = metadataContext{
+		protocols: make(map[multicodec.Code]func() Protocol),
+	}
+	Default.protocols[multicodec.TransportBitswap] = func() Protocol { return &Bitswap{} }
+	Default.protocols[multicodec.TransportGraphsyncFilecoinv1] = func() Protocol { return &GraphsyncFilecoinV1{} }
+}
+
+// WithProtocol dervies a new MetadataContext including the additional protocol mapping.
+func (mc *metadataContext) WithProtocol(id multicodec.Code, factory func() Protocol) metadataContext {
```

Comment:
* Unused?
* Exported function with unexported return type?

Reply:
the 'metadataContext' unexported  can still be used in the same way a `.Default` - it's hidden to prevent a client from making a malformed one directly